app/scripts/background.js

```diff
@@ -435,12 +435,16 @@ function setupController(initState, initLangCode) {
     METAMASK_CONTROLLER_EVENTS.UPDATE_BADGE,
     updateBadge,
   );
-  controller.approvalController.subscribe(updateBadge);
   controller.appStateController.on(
     METAMASK_CONTROLLER_EVENTS.UPDATE_BADGE,
     updateBadge,
   );
 
+  controller.controllerMessenger.subscribe(
+    METAMASK_CONTROLLER_EVENTS.APPROVAL_STATE_CHANGE,
+    updateBadge,
+  );
```

Comment:
somewhat related nit: we never clean up any of these listeners in this file. 

this main controller only gets used once in the lifecycle so we dont need to, but just wanted to point it out

Reply:
I'm pretty confident that we do have event-related leaks, so it's great to keep an eye out for it!